tags/tag_set.go

```diff
@@ -63,26 +71,31 @@ func (ts *TagSet) String() string {
 	return buffer.String()
 }
 
-func (ts *TagSet) insertBytes(k Key, b []byte) bool {
-	if _, ok := ts.m[k]; ok {
-		return false
+// FromContext returns the TagSet stored in the context. The returned TagSet
+// shouldn't be mutated.
+func FromContext(ctx context.Context) *TagSet {
+	ts := ctx.Value(tagSetCtxKey)
+	if ts == nil {
+		return newTagSet(0)
 	}
-	ts.m[k] = b
-	return true
+	return ts.(*TagSet)
 }
 
-func (ts *TagSet) updateBytes(k Key, b []byte) bool {
-	if _, ok := ts.m[k]; !ok {
-		return false
-	}
-	ts.m[k] = b
-	return true
-}
+// TODO(jbd): It says "The returned TagSet shouldn't be mutated.",
+// but tag set cannot be mutated. Remove the comment.
 
-func (ts *TagSet) upsertBytes(k Key, b []byte) {
```

Comment:
I think you should keep these methods. Having the tagSet as a map[k]=v is just for the initial release. But this is a big performance bottleneck (when extracting the keys to project them on the views).
For next release (at least that was my plan) you'll probably need to create your own data structure to encode the (key,value) set.
Upsert, Insert, Update will be quite a bit more complicated but more efficient when extracting the data.

Reply:
Needs to be addressed when performance improvement is being implemented.